pkg/runner/test_runner.go

```diff
@@ -152,68 +153,187 @@ func newTargetWriter(log *logrus.Entry, timeouts TestRunnerTimeouts) *targetWrit
 	return &targetWriter{log: log, timeouts: timeouts}
 }
 
+func (tr *TestRunner) injectTargets(terminate <-chan struct{}, targets []*target.Target, writeChannel chan<- *target.Target, log *logrus.Entry) {
```

Comment:
Just a thought:

```
terminate <-chan struct{}, ..., log *logrus.Entry
```
This is early stages of the problem I'm talking about here: https://fb.quip.com/PUndAbbUFp2p
Since the document might be closed for some users, I'll write TL;DR here:
Basically the thought is that all those loggers, done-contexts (`terminate` in this case), metrics, tracing and other things evolves to a set of argument to each called function through almost any execution flow. And each of them requires manual support each time (to add fields to logger, tags to metrics, recursively create a done-context and so on). So may be in long term we should consider using of smart contexts instead.